### PR TITLE
[3.6] Device state service: fix device not being timeouted bug, fix flaky tests, add more test coverage

### PR DESCRIPTION
**Pull request by dskarzh**
## Pull Request description

[Bug](https://github.com/thingsboard/thingsboard/issues/9156) - wrong timeout expiry check

No documentation changes required

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [ ] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [x] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [x] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [x] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.
  
## Front-End feature checklist

- [ ] Screenshots with affected component(s) are added. The best option is to provide 2 screens: before and after changes;
- [ ] If you change the widget or other API, ensure it is backward-compatible or upgrade script is present.
- [ ] Ensure new API is documented [here](https://github.com/thingsboard/thingsboard-ui-help)

## Back-End feature checklist

- [x] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [x] If new dependency was added: the dependency tree is checked for conflicts.
- [x] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [x] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.

### CODE REVIEW

File: application/src/main/java/org/thingsboard/server/service/state/DefaultDeviceStateService.java [hunk 1]
```diff
@@ -486,7 +486,7 @@ void updateInactivityStateIfExpired(long ts, DeviceId deviceId, DeviceStateData
         if (stateData != null) {
             DeviceState state = stateData.getState();
             if (!isActive(ts, state)
-                    && (state.getLastInactivityAlarmTime() == 0L || state.getLastInactivityAlarmTime() < state.getLastActivityTime())
+                    && (state.getLastInactivityAlarmTime() == 0L || state.getLastInactivityAlarmTime() <= state.getLastActivityTime())
```
Comment (smatvienko-tb) [inline application/src/main/java/org/thingsboard/server/service/state/DefaultDeviceStateService.java:489]: please, add parametrized tests to cover all corner cases

File: application/src/test/java/org/thingsboard/server/service/state/DefaultDeviceStateServiceTest.java [hunk 2]
```diff
@@ -294,6 +294,7 @@ public void decreaseInactivityForActiveDeviceTest() throws Exception {
         Mockito.verify(telemetrySubscriptionService, Mockito.never()).saveAttrAndNotify(Mockito.any(), Mockito.eq(deviceId), Mockito.any(), Mockito.eq("active"), Mockito.any(), Mockito.any());
 
         long newTimeout = 1;
+        Thread.sleep(newTimeout);
```
Comment (smatvienko-tb) [inline application/src/test/java/org/thingsboard/server/service/state/DefaultDeviceStateServiceTest.java:297]: This test fails 100% in the same millisecond. As we saw when we mocked the system time wrapper.

I agree with your suggestion to add another test. But in the parametrized way.
AtomicLong atomicTs = new AtomicLong(); // in the Setup section

Lets mock `willAnswer (() -> atomicTs.addAndGet(msFromParameter)).given(service).currentTimeMs()`
 msFromParameter=0 - you got a frozen time
 msFromParameter=1 - you got controlled time each call will return the next ms (this is default mock)

**Review by smatvienko-tb — CHANGES_REQUESTED** [review]
please, find my comments below

**Comment by dskarzh** [comment]
What has changed:
- Migrated to JUnit 5 for better support for parameterized tests
- Minor code style changes like adding static imports to avoid repeating Mockito each time, replaced literals with constants etc..
- Removed redundant `Thread.sleep()` from `decreaseInactivityForInactiveDeviceTest()`. Reason why it initially was added turned out to affect only `decreaseInactivityForActiveDeviceTest()` 
- Added test to with controlled time for decreasing inactivity timeout use case when device is active: now correctly checks that if timeout is decreased to minimum value (1) and it has not yet passed then device should stay active
- Added test to cover normal use cases and corner cases of `updateInactivityStateIfExpired()`. This test is not in any way exhaustive and was designed to check only realistic conditions. If we account for user having full control over some attributes it becomes very hard to test all cases because time can be negative etc. (Maybe it is a good idea to restrict user from changing attributes used to represent device state and giving him alternative ways to solve use cases which previously were solved by directly manipulating device state attributes. Other side of this is that restricting direct access to these attributes is a breaking change, because a lot of people rely on it)
- Made `stateData.getDeviceCreationTime() + state.getInactivityTimeout() = ts` check to be constistent with `isActive()`.

**Comment by dskarzh** [comment]
image
